# Optimizing Data Acquisition

Since this is a pretty extensive dataset, I'll explore jsut one subset of the data. Let's see what can be done to reduce size

In [1]:
import pandas as pd
ny = pd.read_csv(r'..\data\raw\taxi data\yellow_tripdata_2015-01_00', parse_dates=['tpep_pickup_datetime', 'tpep_dropoff_datetime'])
ny.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2015-01-15 19:05:39   2015-01-15 19:23:42                1   
1         1  2015-01-10 20:33:38   2015-01-10 20:53:28                1   
2         1  2015-01-10 20:33:38   2015-01-10 20:43:41                1   
3         1  2015-01-10 20:33:39   2015-01-10 20:35:31                1   
4         1  2015-01-10 20:33:39   2015-01-10 20:52:58                1   

   trip_distance  pickup_longitude  pickup_latitude  RateCodeID  \
0           1.59        -73.993896        40.750111           1   
1           3.30        -74.001648        40.724243           1   
2           1.80        -73.963341        40.802788           1   
3           0.50        -74.009087        40.713818           1   
4           3.00        -73.971176        40.762428           1   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude  payment_type  \
0                  N         -73.974785         40.750618             1   
1                  N         -73.994415         40.759109             1   
2                  N         -73.951820         40.824413             2   
3                  N         -74.004326         40.719986             2   
4                  N         -74.004181         40.742653             2   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0         12.0    1.0      0.5        3.25           0.0   
1         14.5    0.5      0.5        2.00           0.0   
2          9.5    0.5      0.5        0.00           0.0   
3          3.5    0.5      0.5        0.00           0.0   
4         15.0    0.5      0.5        0.00           0.0   

   improvement_surcharge  total_amount  
0                    0.3         17.05  
1                    0.3         17.80  
2                    0.3         10.80  
3                    0.3          4.80  
4                    0.3         16.30

In [2]:
ny.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499999 entries, 0 to 499998
Data columns (total 19 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               499999 non-null  int64         
 1   tpep_pickup_datetime   499999 non-null  datetime64[ns]
 2   tpep_dropoff_datetime  499999 non-null  datetime64[ns]
 3   passenger_count        499999 non-null  int64         
 4   trip_distance          499999 non-null  float64       
 5   pickup_longitude       499999 non-null  float64       
 6   pickup_latitude        499999 non-null  float64       
 7   RateCodeID             499999 non-null  int64         
 8   store_and_fwd_flag     499999 non-null  object        
 9   dropoff_longitude      499999 non-null  float64       
 10  dropoff_latitude       499999 non-null  float64       
 11  payment_type           499999 non-null  int64         
 12  fare_amount            499999 non-null  floa

I'll try to reduce the footprint by selecting the data types used for each field.

In [3]:
dtype = {
    'VendorID': 'category',
    'passenger_count': 'int8',
    'trip_distance': 'float32',
    'pickup_longitude': 'float32',
    'pickup_latitude': 'float32',
    'dropoff_longitude': 'float32',
    'dropoff_latitude': 'float32',
    'RateCodeID': 'category',
    'store_and_fwd_flag': 'category',
    'payment_type': 'category',
    'fare_amount': 'float32',
    'extra': 'float32',
    'mta_tax': 'category',
    'tip_amount': 'float32',
    'tolls_amount': 'float32',
    'improvement_surcharge': 'category',
    'total_amount': 'float32'
}

ny = pd.read_csv(r'..\data\raw\taxi data\yellow_tripdata_2015-01_00', dtype=dtype, parse_dates=['tpep_pickup_datetime', 'tpep_dropoff_datetime'])
ny.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499999 entries, 0 to 499998
Data columns (total 19 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               499999 non-null  category      
 1   tpep_pickup_datetime   499999 non-null  datetime64[ns]
 2   tpep_dropoff_datetime  499999 non-null  datetime64[ns]
 3   passenger_count        499999 non-null  int8          
 4   trip_distance          499999 non-null  float32       
 5   pickup_longitude       499999 non-null  float32       
 6   pickup_latitude        499999 non-null  float32       
 7   RateCodeID             499999 non-null  category      
 8   store_and_fwd_flag     499999 non-null  category      
 9   dropoff_longitude      499999 non-null  float32       
 10  dropoff_latitude       499999 non-null  float32       
 11  payment_type           499999 non-null  category      
 12  fare_amount            499999 non-null  floa

We have gone from 72.5 MB to to 28.1 MB.

# Cleaning Data

Now we'll try to find inconsistencies in the data. Let's look at its description first.

In [4]:
ny.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0        2  2015-01-15 19:05:39   2015-01-15 19:23:42                1   
1        1  2015-01-10 20:33:38   2015-01-10 20:53:28                1   
2        1  2015-01-10 20:33:38   2015-01-10 20:43:41                1   
3        1  2015-01-10 20:33:39   2015-01-10 20:35:31                1   
4        1  2015-01-10 20:33:39   2015-01-10 20:52:58                1   

   trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
0           1.59        -73.993896        40.750111          1   
1           3.30        -74.001648        40.724243          1   
2           1.80        -73.963341        40.802788          1   
3           0.50        -74.009087        40.713818          1   
4           3.00        -73.971176        40.762428          1   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude payment_type  \
0                  N         -73.974785         40.750618            1   
1                  N         -73.994415         40.759109            1   
2                  N         -73.951820         40.824413            2   
3                  N         -74.004326         40.719986            2   
4                  N         -74.004181         40.742653            2   

   fare_amount  extra mta_tax  tip_amount  tolls_amount improvement_surcharge  \
0         12.0    1.0     0.5        3.25           0.0                   0.3   
1         14.5    0.5     0.5        2.00           0.0                   0.3   
2          9.5    0.5     0.5        0.00           0.0                   0.3   
3          3.5    0.5     0.5        0.00           0.0                   0.3   
4         15.0    0.5     0.5        0.00           0.0                   0.3   

   total_amount  
0     17.049999  
1     17.799999  
2     10.800000  
3      4.800000  
4     16.299999

In [5]:
print(ny.shape)
ny.describe()

(499999, 19)


passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count    499999.000000  499999.000000     499999.000000    499999.000000   
mean          1.678109       3.019222        -72.560242        39.971840   
std           1.334983     136.926987         10.130702         5.581226   
min           0.000000       0.000000        -87.451874         0.000000   
25%           1.000000       1.000000        -73.991669        40.735725   
50%           1.000000       1.660000        -73.981567        40.753300   
75%           2.000000       3.000000        -73.966606        40.767715   
max           9.000000   92000.898438          0.000000        42.294155   

       dropoff_longitude  dropoff_latitude    fare_amount          extra  \
count      499999.000000     499999.000000  499999.000000  499999.000000   
mean          -72.605782         39.998131      11.872335       0.314752   
std             9.968216          5.491283      10.133603       0.367731   
min           -86.731705          0.000000    -138.899994      -1.000000   
25%           -73.991219         40.734528       6.500000       0.000000   
50%           -73.979820         40.753757       9.000000       0.000000   
75%           -73.962524         40.768940      13.500000       0.500000   
max             0.000000         49.194656     900.000000       7.000000   

          tip_amount   tolls_amount   total_amount  
count  499999.000000  499999.000000  499999.000000  
mean        1.543074       0.241704      14.768952  
std         2.511312       1.222974      12.300300  
min       -81.000000      -5.330000    -139.699997  
25%         0.000000       0.000000       8.160000  
50%         1.000000       0.000000      11.150000  
75%         2.060000       0.000000      16.299999  
max       750.000000      95.330002     900.299988

It seems there are negative values in values related to charges (total_amount, extra...).

In [6]:
idx = ny['total_amount'] < 0
print(ny[idx].shape)
ny[idx].head(10)

(158, 19)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
3831         2  2015-01-17 22:40:27   2015-01-17 22:43:04                1   
4924         2  2015-01-15 17:33:24   2015-01-15 17:33:31                2   
10046        2  2015-01-16 16:00:45   2015-01-16 16:00:53                1   
16703        2  2015-01-31 23:38:52   2015-01-31 23:38:54                2   
19953        2  2015-01-10 02:23:53   2015-01-10 02:23:58                2   
22989        2  2015-01-14 11:52:09   2015-01-14 11:52:20                1   
23879        2  2015-01-03 02:01:25   2015-01-03 02:01:54                1   
26582        2  2015-01-12 15:07:29   2015-01-12 15:07:35                1   
27172        2  2015-01-06 14:07:25   2015-01-06 14:08:27                1   
30188        2  2015-01-10 21:10:20   2015-01-10 21:12:39                1   

       trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
3831            0.11        -74.002357        40.739826          1   
4924            0.00        -73.982567        40.739799          1   
10046           0.00        -73.937721        40.758194          1   
16703           0.00          0.000000         0.000000          2   
19953           0.00          0.000000         0.000000          5   
22989           0.00        -73.789955        40.646946          2   
23879           0.03        -73.953407        40.811150          1   
26582           0.00          0.000000         0.000000          2   
27172           0.03        -73.994560        40.740318          1   
30188           0.03        -73.986328        40.755280          1   

      store_and_fwd_flag  dropoff_longitude  dropoff_latitude payment_type  \
3831                   N         -74.001114         40.741108            4   
4924                   N         -73.982567         40.739799            3   
10046                  N         -73.937721         40.758194            3   
16703                  N           0.000000          0.000000            2   
19953                  N           0.000000          0.000000            1   
22989                  N           0.000000          0.000000            3   
23879                  N         -73.953751         40.811302            2   
26582                  N           0.000000          0.000000            2   
27172                  N         -73.995331         40.740952            4   
30188                  N         -73.985420         40.755089            4   

       fare_amount  extra mta_tax  tip_amount  tolls_amount  \
3831          -3.5   -0.5    -0.5        0.00          0.00   
4924          -2.5   -1.0    -0.5       -0.70          0.00   
10046         -2.5   -1.0    -0.5        0.00          0.00   
16703        -52.0    0.0    -0.5        0.00          0.00   
19953         -6.8    0.0       0       -1.00          0.00   
22989        -52.0    0.0    -0.5      -14.33         -5.33   
23879         -2.5   -0.5    -0.5        0.00          0.00   
26582        -52.0    0.0    -0.5        0.00          0.00   
27172         -2.5    0.0    -0.5        0.00          0.00   
30188         -3.5   -0.5    -0.5        0.00          0.00   

      improvement_surcharge  total_amount  
3831                    0.3     -4.800000  
4924                    0.3     -5.000000  
10046                   0.3     -4.300000  
16703                   0.3    -52.799999  
19953                   0.3     -8.100000  
22989                   0.3    -72.459999  
23879                   0.3     -3.800000  
26582                   0.3    -52.799999  
27172                   0.3     -3.300000  
30188                   0.3     -4.800000

It doesn't make sense unless they are refunds. In that case, it means there are 2 records for the same pickup, one for the original trip and one for the refund.

The data dictionary [https://storage.googleapis.com/hiring-test/data_dictionary_trip_records_yellow.pdf](link) defines the following categories of payment in the payment_type variable:

 - 1= Credit card 
 - 2= Cash
 - 3= No charge 
 - 4= Dispute
 - 5= Unknown 
 - 6= Voided trip

But from the small sample above it seems clear not all refunds are registered as 'Dispute', 'Voided trip'... So this field can't be trusted to identify refunds.

Moving forward, records with a negative total_amount should be discarded to avoid duplicity.


In [7]:
ny = ny[ny['total_amount'] >= 0]
print(ny.shape)
ny.describe()

(499841, 19)


passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count    499841.000000  499841.000000     499841.000000    499841.000000   
mean          1.678048       3.020100        -72.563202        39.973476   
std           1.334955     136.948608         10.120315         5.575504   
min           0.000000       0.000000        -87.451874         0.000000   
25%           1.000000       1.000000        -73.991669        40.735729   
50%           1.000000       1.660000        -73.981567        40.753304   
75%           2.000000       3.000000        -73.966614        40.767715   
max           9.000000   92000.898438          0.000000        42.294155   

       dropoff_longitude  dropoff_latitude    fare_amount          extra  \
count      499841.000000     499841.000000  499841.000000  499841.000000   
mean          -72.609505         40.000172      11.879289       0.314949   
std             9.954946          5.483973      10.122419       0.367572   
min           -86.731705          0.000000       0.000000       0.000000   
25%           -73.991219         40.734531       6.500000       0.000000   
50%           -73.979820         40.753761       9.000000       0.000000   
75%           -73.962532         40.768944      13.500000       0.500000   
max             0.000000         49.194656     900.000000       7.000000   

          tip_amount   tolls_amount   total_amount  
count  499841.000000  499841.000000  499841.000000  
mean        1.543882       0.241802      14.777504  
std         2.507314       1.223109      12.287228  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       8.160000  
50%         1.000000       0.000000      11.150000  
75%         2.060000       0.000000      16.299999  
max       750.000000      95.330002     900.299988

It also stands out that the minimum latitude for the pickups and the dropoffs is 0 (pretty far from NYC). The same happens with the max longitude.

In [8]:
idx = ny['pickup_longitude'] == 0
print(ny[idx].shape)
ny[idx].head()

(9537, 19)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
31         2  2015-01-15 19:05:43   2015-01-15 19:05:44                2   
61         1  2015-01-04 13:44:52   2015-01-04 13:56:49                1   
66         2  2015-01-04 13:44:52   2015-01-04 13:49:03                1   
157        1  2015-01-15 09:47:00   2015-01-15 10:00:07                1   
159        1  2015-01-15 09:47:02   2015-01-15 10:17:47                3   

     trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
31            0.01               0.0              0.0          5   
61            2.50               0.0              0.0          1   
66            0.85               0.0              0.0          1   
157           1.00               0.0              0.0          1   
159           8.30               0.0              0.0          1   

    store_and_fwd_flag  dropoff_longitude  dropoff_latitude payment_type  \
31                   N                0.0               0.0            1   
61                   N                0.0               0.0            1   
66                   N                0.0               0.0            2   
157                  N                0.0               0.0            2   
159                  N                0.0               0.0            1   

     fare_amount  extra mta_tax  tip_amount  tolls_amount  \
31          60.0    0.0       0        0.00          0.00   
61          11.0    0.0     0.5        2.35          0.00   
66           5.5    0.0     0.5        0.00          0.00   
157         10.0    0.0     0.5        0.00          0.00   
159         27.5    0.0     0.5       10.00          5.33   

    improvement_surcharge  total_amount  
31                    0.3     60.299999  
61                      0     14.150000  
66                    0.3      6.300000  
157                   0.3     10.800000  
159                   0.3     43.630001

In [9]:
ny = ny[ny['trip_distance'] != 0]
print(ny.shape)
ny.describe()

(496622, 19)


passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count    496622.000000  496622.000000     496622.000000    496622.000000   
mean          1.680282       3.039675        -72.656624        40.024918   
std           1.336804     137.391525          9.786289         5.391513   
min           0.000000       0.010000        -87.451874         0.000000   
25%           1.000000       1.000000        -73.991684        40.735886   
50%           1.000000       1.680000        -73.981606        40.753380   
75%           2.000000       3.000000        -73.966766        40.767754   
max           6.000000   92000.898438          0.000000        42.294155   

       dropoff_longitude  dropoff_latitude    fare_amount          extra  \
count      496622.000000     496622.000000  496622.000000  496622.000000   
mean          -72.779053         40.093559      11.838174       0.315709   
std             9.327368          5.138230       9.739705       0.367710   
min           -86.731705          0.000000       0.000000       0.000000   
25%           -73.991241         40.734730       6.500000       0.000000   
50%           -73.979897         40.753876       9.000000       0.000000   
75%           -73.962791         40.769020      13.500000       0.500000   
max             0.000000         49.194656     420.000000       7.000000   

          tip_amount   tolls_amount   total_amount  
count  496622.000000  496622.000000  496622.000000  
mean        1.538152       0.240555      14.730853  
std         2.225962       1.218816      11.885249  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       8.300000  
50%         1.000000       0.000000      11.160000  
75%         2.060000       0.000000      16.299999  
max       200.000000      95.330002     453.299988

It's clear that the location data is invalid for these records. This data will get discarded later when joining with the geojson data. For now I'll discard manually

In [10]:
ny = ny[ny['pickup_longitude'] != 0]
ny = ny[ny['dropoff_longitude'] != 0]
print(ny.shape)
ny.describe()

(487557, 19)


passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count    487557.000000  487557.000000     487557.000000    487557.000000   
mean          1.685159       3.045639        -73.974747        40.751041   
std           1.342100     138.662109          0.043394         0.075289   
min           0.000000       0.010000        -87.451874         4.789132   
25%           1.000000       1.000000        -73.991859        40.737282   
50%           1.000000       1.680000        -73.981918        40.754162   
75%           2.000000       3.000000        -73.967941        40.768108   
max           6.000000   92000.898438        -69.598526        42.294155   

       dropoff_longitude  dropoff_latitude    fare_amount          extra  \
count      487557.000000     487557.000000  487557.000000  487557.000000   
mean          -73.974266         40.751987      11.834809       0.315770   
std             0.115499          0.048327       9.712847       0.367778   
min           -86.731705         18.625944       0.000000       0.000000   
25%           -73.991379         40.736206       6.500000       0.000000   
50%           -73.980286         40.754509       9.000000       0.000000   
75%           -73.964012         40.769390      13.500000       0.500000   
max            -0.116667         49.194656     420.000000       7.000000   

          tip_amount   tolls_amount   total_amount  
count  487557.000000  487557.000000  487557.000000  
mean        1.536997       0.240602      14.726496  
std         2.217087       1.219360      11.857430  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       8.300000  
50%         1.000000       0.000000      11.160000  
75%         2.060000       0.000000      16.299999  
max       200.000000      95.330002     453.299988

In [11]:
print(ny[ny['trip_distance'] == 0].shape)
ny[ny['trip_distance'] == 0].head()

(0, 19)


Empty DataFrame
Columns: [VendorID, tpep_pickup_datetime, tpep_dropoff_datetime, passenger_count, trip_distance, pickup_longitude, pickup_latitude, RateCodeID, store_and_fwd_flag, dropoff_longitude, dropoff_latitude, payment_type, fare_amount, extra, mta_tax, tip_amount, tolls_amount, improvement_surcharge, total_amount]
Index: []

Either the pickup or the dropoff is wrong. Or maybe these records are additional charges made at the dropoff location after a wrong mischarge. If that's the case this is not actually a pickup, so I'm inclinded to delete this records (although I recognize an argument could be made towards keeping them).

There still seem to be wrong coordinates (max dropoff longitude = -0.11) but as I said, they will be discarded later.

There are also lines where the trip distance is 0.

Finally, there are rows where the passenger count is 0, but this can be attributed to a wrongful registry of the data and it doesn't really affect the scope of this analysis.

We started with 499999 records and after cleaning we have 487557. In the next step, we'll try to clean the whole dataset.

# Joining with GEOJSON data

We need to calculate the average (we assume daily) pickups by block. Blocks are defined in the nyc_cbg_geoms.geoson file. GeoPandas allow us to easily join coordinates with geojson polygon data.

In [32]:
import geopandas

nyg = geopandas.GeoDataFrame(
    ny, 
    geometry=geopandas.points_from_xy(
        ny['pickup_longitude'], ny['pickup_latitude'], crs=4326)
)
nyg.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0        2  2015-01-15 19:05:39   2015-01-15 19:23:42                1   
1        1  2015-01-10 20:33:38   2015-01-10 20:53:28                1   
2        1  2015-01-10 20:33:38   2015-01-10 20:43:41                1   
3        1  2015-01-10 20:33:39   2015-01-10 20:35:31                1   
4        1  2015-01-10 20:33:39   2015-01-10 20:52:58                1   

   trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
0           1.59        -73.993896        40.750111          1   
1           3.30        -74.001648        40.724243          1   
2           1.80        -73.963341        40.802788          1   
3           0.50        -74.009087        40.713818          1   
4           3.00        -73.971176        40.762428          1   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude payment_type  \
0                  N         -73.974785         40.750618            1   
1                  N         -73.994415         40.759109            1   
2                  N         -73.951820         40.824413            2   
3                  N         -74.004326         40.719986            2   
4                  N         -74.004181         40.742653            2   

   fare_amount  extra mta_tax  tip_amount  tolls_amount improvement_surcharge  \
0         12.0    1.0     0.5        3.25           0.0                   0.3   
1         14.5    0.5     0.5        2.00           0.0                   0.3   
2          9.5    0.5     0.5        0.00           0.0                   0.3   
3          3.5    0.5     0.5        0.00           0.0                   0.3   
4         15.0    0.5     0.5        0.00           0.0                   0.3   

   total_amount                    geometry  
0     17.049999  POINT (-73.99390 40.75011)  
1     17.799999  POINT (-74.00165 40.72424)  
2     10.800000  POINT (-73.96334 40.80279)  
3      4.800000  POINT (-74.00909 40.71382)  
4     16.299999  POINT (-73.97118 40.76243)

In [13]:
geo = geopandas.read_file(r'..\data\raw\nyc_cbg_geoms.geojson')
geo.head()

geoid                                           geometry
0  360050001001  POLYGON ((-73.89277 40.79284, -73.89261 40.792...
1  360050002001  POLYGON ((-73.86285 40.81267, -73.86191 40.812...
2  360050002002  POLYGON ((-73.86708 40.81444, -73.86332 40.812...
3  360050002003  POLYGON ((-73.85856 40.80665, -73.85848 40.806...
4  360050004001  POLYGON ((-73.85972 40.81527, -73.85956 40.815...

In [14]:
%%time
nyg_join = geopandas.sjoin(nyg, geo, how='inner', op='within')

Wall time: 38.3 s


In [15]:
print(nyg_join.shape)
nyg_join.head()

(487202, 22)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2015-01-15 19:05:39   2015-01-15 19:23:42                1   
11        1  2015-01-10 20:33:41   2015-01-10 20:43:26                1   
26        2  2015-01-15 19:05:42   2015-01-15 19:16:18                1   
64        2  2015-01-04 13:44:52   2015-01-04 13:53:17                1   
89        2  2015-01-15 14:00:45   2015-01-15 14:10:24                1   

    trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
0            1.59        -73.993896        40.750111          1   
11           1.10        -73.993782        40.751419          1   
26           1.53        -73.991127        40.750080          1   
64           1.14        -73.990952        40.750809          1   
89           1.20        -73.994003        40.751350          1   

   store_and_fwd_flag  dropoff_longitude  ...  fare_amount extra  mta_tax  \
0                   N         -73.974785  ...         12.0   1.0      0.5   
11                  N         -73.967407  ...          7.5   0.5      0.5   
26                  N         -73.988609  ...          9.0   1.0      0.5   
64                  N         -73.979408  ...          7.0   0.0      0.5   
89                  N         -73.981354  ...          8.5   0.0      0.5   

    tip_amount tolls_amount  improvement_surcharge  total_amount  \
0         3.25          0.0                    0.3     17.049999   
11        1.00          0.0                    0.3      9.800000   
26        0.00          0.0                    0.3     10.800000   
64        1.40          0.0                    0.3      9.200000   
89        0.00          0.0                    0.3      9.300000   

                      geometry  index_right         geoid  
0   POINT (-73.99390 40.75011)         3575  360610101001  
11  POINT (-73.99378 40.75142)         3575  360610101001  
26  POINT (-73.99113 40.75008)         3575  360610101001  
64  POINT (-73.99095 40.75081)         3575  360610101001  
89  POINT (-73.99400 40.75135)         3575  360610101001  

[5 rows x 22 columns]

Since the spatial join is very time consuming and we only need the average number of pickups, we'll calculate from the provided partitioned data.

In [16]:
nyg_join['pickup_date'] = nyg_join['tpep_pickup_datetime'].dt.floor('d')
nyg_join.groupby(['pickup_date', 'geoid']) \
    .size().reset_index(name='avg') \
    .groupby('geoid') \
    .mean()

avg
geoid                 
360050001001  2.000000
360050004002  1.000000
360050016004  1.000000
360050019001  1.642857
360050019002  1.000000
...                ...
360850169012  1.000000
360850177003  1.000000
360850207001  1.000000
360850319011  1.000000
361190015031  1.000000

[2585 rows x 1 columns]

In [17]:
del nyg_join

# Using PostGIS

The ~40 seconds it takes to complete the spatial join multiplied by 75 files means almost an hour of computing time. We probably could use some alternative. There are resources online that claim faster spatial joins using PostGIS, so I created a test database to do some benchmarks.

In [18]:
from sqlalchemy import create_engine
from geoalchemy2 import Geometry, WKTElement

engine = create_engine('postgresql://postgres:1234@localhost:5432/carto')

In [34]:
%%time
nyg_min = nyg[['geometry', 'tpep_pickup_datetime']]
nyg_min.to_postgis('trip', engine, if_exists='append', index=False)

Wall time: 30.4 s


In [26]:
%%time
geo.to_postgis('block', engine)

Wall time: 925 ms


In [39]:
%%time
sql =  """SELECT COUNT(*) FROM trip INNER JOIN block 
            ON ST_Within(trip.geometry, block.geometry) 
            GROUP BY geoid, date_trunc('day', trip.tpep_pickup_datetime)"""
pd.read_sql_query(sql, engine)

Wall time: 13 s


It seems PostGIS makes spatial joins a lot faster than GeoPandas. Unfortunately, uploading the data from GeoPandas is also very slow. I'll try to upload using PostGres Copy command to see if it is worth it.

This time I'll try to upload all the columns. To create the table I'll take advantage of the to_postgis method.

In [63]:
import numpy as np
engine.execute('DROP TABLE trip')
nyg[np.full(nyg.shape[0], False)].to_postgis('trip', engine, if_exists='append', index=False)
engine.execute('ALTER TABLE trip DROP COLUMN geometry')